### PR TITLE
Allow modifiers like ChordSymbol to be attached to GlyphNotes.

### PR DESCRIPTION
**Pull request by AaronDavidNewman**
This change allows modifiers to be attached to GlyphNotes.  This is used to notate chord changes in an improvised solo.

image

### CODE REVIEW

**Comment by AaronDavidNewman** [comment]
Here is the test case I added:

image

File: src/glyphnote.ts [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
+import { ModifierContextState } from './modifiercontext';
+
 export interface GlyphNoteOptions {
   ignoreTicks?: boolean;
   line?: number;
 }
 
 export class GlyphNote extends Note {
   protected options: GlyphNoteOptions;
+  static format(notes: GlyphNote[], state: ModifierContextState): boolean {
```
Comment (AaronDavidNewman) [inline src/glyphnote.ts:17]: Do I need to do this?  I am following the pattern of StaveNote, and I needed to do this to get the format() of the modifiers called.  But maybe there is a way I can add the modifiers and not the GlyphNote itself.

Comment (AaronDavidNewman) [reply src/glyphnote.ts]: Indeed I do not, I can just format the modifiers and not the thing itself.  Removed the empty call.

File: src/glyphnote.ts [hunk 2]
```diff
@@ -1,15 +1,20 @@
 // [VexFlow](http://vexflow.com) - Copyright (c) Mohit Muthanna 2010.
 
+import { Vex } from './vex';
 import { BoundingBox } from './boundingbox';
 import { Glyph } from './glyph';
 import { Note, NoteStruct } from './note';
+import { ModifierContext } from './modifiercontext';
+import { ModifierContextState } from './modifiercontext';
```
Comment (rvilarl) [inline src/glyphnote.ts:8]: This line can be deleted, `ModifierContextState` is not used.

File: src/glyphnote.ts [hunk 3]
```diff
@@ -36,31 +41,54 @@ export class GlyphNote extends Note {
...
+    for (let i = 0; i < this.modifiers.length; ++i) {
+      this.modifierContext.addMember(this.modifiers[i]);
+    }
+    this.setPreFormatted(false);
     return this;
   }
-  */
 
   preFormat(): this {
+    if (!this.preFormatted && this.modifierContext) {
```
Comment (rvilarl) [inline src/glyphnote.ts:56]: Should an error be thrown if `this.modifierContext` is `undefined`?

File: src/glyphnote.ts [hunk 4]
```diff
@@ -36,31 +41,54 @@ export class GlyphNote extends Note {
...
+    ctx.closeGroup();
+  }
   draw(): void {
-    const stave = this.checkStave();
-    const ctx = stave.checkContext();
+    if (!this.stave) {
+      throw new Vex.RERR('NoStave', 'No stave attached to this note.');
+    }
+
+    const ctx = this.stave.checkContext();
```
Comment (rvilarl) [inline src/glyphnote.ts:77]: `this.checkStave()` works with `this.stave` like `this.checkContext()` with `this.context`:
- `this.checkStave()` will throw an `RERR` if `this.stave` is `undefined`. 

Please revert the change.

File: src/modifiercontext.ts [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
 import { Vex } from './vex';
 import { StaveNote } from './stavenote';
+import { GlyphNote } from './glyphnote';
```
Comment (rvilarl) [inline src/modifiercontext.ts:10]: This line can be deleted. `GlyphNote` is not used.

**Review by rvilarl — CHANGES_REQUESTED** [review]

File: src/glyphnote.ts (hunk 3 above)
Comment (AaronDavidNewman) [reply src/glyphnote.ts:56]: I don't know.  I followed the lead from stavenote.ts, it allows modifierContext to be null in preFormat.

File: src/glyphnote.ts (hunk 2 above)
Comment (AaronDavidNewman) [reply src/glyphnote.ts]: done.  Didn't it used to throw an error when you did that?

File: src/modifiercontext.ts (hunk 5 above)
Comment (AaronDavidNewman) [reply src/modifiercontext.ts]: done

File: src/glyphnote.ts (hunk 4 above)
Comment (AaronDavidNewman) [reply src/glyphnote.ts]: Sorry, that must have snuck in from my other branch.  I reverted to what it was.

File: src/glyphnote.ts [hunk 6]
```diff
@@ -1,15 +1,19 @@
 // [VexFlow](http://vexflow.com) - Copyright (c) Mohit Muthanna 2010.
 
+import { Vex } from './vex';
```
Comment (0xfe) [inline src/glyphnote.ts:3]: This line can be removed. (unused Vex).

File: src/glyphnote.ts [hunk 7]
```diff
@@ -36,21 +40,39 @@ export class GlyphNote extends Note {
     return this.glyph.getBoundingBox();
   }
 
-  /*
-  addToModifierContext() {
+  // Add self to modifier context. `mContext` is the `ModifierContext`
+  // to be added to.
+  addToModifierContext(mContext: ModifierContext): this {
```
Comment (0xfe) [inline src/glyphnote.ts:45]: Is this necessary? Isn't it derived from the inheritance chain?

File: src/glyphnote.ts (hunk 3 above)
Comment (0xfe) [reply src/glyphnote.ts:55]: Right, this is okay.

File: src/glyphnote.ts [hunk 8]
```diff
@@ -36,21 +40,39 @@ export class GlyphNote extends Note {
...
+  drawModifiers(): void {
+    const ctx = this.checkContext();
+    ctx.openGroup('modifiers');
+    for (let i = 0; i < this.modifiers.length; i++) {
+      const modifier = this.modifiers[i];
+      modifier.setContext(ctx);
+      modifier.drawWithStyle();
+    }
+    ctx.closeGroup();
+  }
```
Comment (0xfe) [inline src/glyphnote.ts:70]: New line between method bodies please.

**Review by 0xfe — COMMENTED** [review]
Looks good. Minor comments.

File: src/glyphnote.ts (hunk 7 above)
Comment (AaronDavidNewman) [reply src/glyphnote.ts:45]: The inherited one (`tickable`) doesn't add modifiers to the context. 

We could add this loop code to `Note`, then `stavenote` and `glyphnote` could share, but we'll have to check all the other `Note` children.  We could modify the `tickable` code to check for modifiers, maybe.

Comment (AaronDavidNewman) [reply src/glyphnote.ts:45]: Welp, not exactly the same.  Glyphnote doesn't need to `format`, so it doesn't add itself to the context.  So stavenote looks like this, not sure how you feel about `super`.

``` javascript
    this.modifierContext.addMember(this);
    return super.addToModifierContext(mContext);
```

Comment (AaronDavidNewman) [reply src/glyphnote.ts:45]: welp, this doesn't work, I'd have to add a check for null to `this.modifierContext` which seems to defeat the purpose.  Or add an empty format method for glyphnote.  

So I think I'm back to:  it is necessary :)

```
ERROR in C:\Users\Daddy\Documents\GitHub\vexflow\src\stavenote.ts
./src/stavenote.ts 871:4-24
```

File: src/glyphnote.ts (hunk 6 above)
Comment (AaronDavidNewman) [reply src/glyphnote.ts]: ok

File: src/glyphnote.ts (hunk 2 above)
Comment (rvilarl) [reply src/glyphnote.ts]: The line is still there

File: src/glyphnote.ts (hunk 4 above)
Comment (rvilarl) [reply src/glyphnote.ts]: This is not reverted, did you push the change?

File: src/modifiercontext.ts (hunk 5 above)
Comment (rvilarl) [reply src/modifiercontext.ts]: The line is still there also

File: src/glyphnote.ts (hunk 7 above)
Comment (0xfe) [reply src/glyphnote.ts:44]: kk, thanks for digging in :-)

**Review by 0xfe — APPROVED** [review]
Looks great.
